### PR TITLE
Better logic to derive vpcRegion/Zone from vpcName/Subnets

### PR DESCRIPTION
**Pull request by miyamotoh**
This change would allow users to specify which region to create a new or locate an existing VPC instance and its subnet. With this, users would no longer be tied to the installer-picked region for VPC resources that is hardcoded to be close to the PowerVS region, essentially allowing reuse of an existing VPC infrastructure the user always has, regardless of where the PowerVS instance is located.

The relevant params for VPC resources in question are; 1) `vpcRegion`, 2) `vpcName`, and 3) `vpcSubnets`.

Whether `publish: External` (default) or `publish: Internal`, when `vpcRegion` is specified, the VPC instance the installer works with will be in that region. If `vpcName` is specified additionally, a VPC instance with that name has to be found in that region. If not, one will be created in that region. 

If `vpcSubnets` is specified additionally, the subnet must be found in and attached to the VPC specified by `vpcName`. Custom `vpcSubnets` without `vpcName` will not be supported and thus error outcl. 

If `vpcName` is provided without `vpcRegion`, the VPC by the name will be searched in the nearby region of the PowerVS region, and would be accepted/reused if found, error out otherwise.

The `validation[_test].go` changes would validate all these combos of `vpc*` params.

Signed-off-by: Hiro Miyamoto <miyamotoh@us.ibm.com>

### CODE REVIEW

File: pkg/asset/cluster/tfvars.go [hunk 1]
```diff
@@ -62,6 +66,7 @@ import (
 	"github.com/openshift/installer/pkg/types/openstack"
 	"github.com/openshift/installer/pkg/types/ovirt"
 	"github.com/openshift/installer/pkg/types/powervs"
+	powervstypes "github.com/openshift/installer/pkg/types/powervs"
```
Comment (barbacbd) [inline pkg/asset/cluster/tfvars.go:69]: This is the same import as the line above. Can we consolidate them?

File: pkg/asset/installconfig/powervs/client.go [hunk 2]
```diff
@@ -515,3 +516,24 @@ func (c *Client) GetAuthenticatorAPIKeyDetails(ctx context.Context) (*iamidentit
 func (c *Client) GetAPIKey() string {
 	return c.APIKey
 }
+
+// GetVPCs gets all VPCs in a region
+func (c *Client) GetVPCs(ctx context.Context, region string) ([]vpcv1.VPC, error) {
+	_, cancel := context.WithTimeout(ctx, 1*time.Minute)
```
Comment (barbacbd) [inline pkg/asset/installconfig/powervs/client.go:522]: The installer recreates the context:
`ctx, cancel := context.WithTimeout(ctx, 1*time.Minute)`

File: pkg/asset/installconfig/powervs/client.go [hunk 3]
```diff
@@ -515,3 +516,24 @@ func (c *Client) GetAuthenticatorAPIKeyDetails(ctx context.Context) (*iamidentit
...
+	err := c.SetVPCServiceURLForRegion(ctx, region)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to set vpc api service url")
+	}
+
+	allVPCs := []vpcv1.VPC{}
+	if vpcs, detailedResponse, err := c.vpcAPI.ListVpcs(c.vpcAPI.NewListVpcsOptions()); err != nil {
+		if detailedResponse.GetStatusCode() != http.StatusNotFound {
+			return nil, err
+		}
```
Comment (barbacbd) [inline pkg/asset/installconfig/powervs/client.go:534]: An error occurred, is there a reason to skip it here ?

Comment (miyamotoh) [reply pkg/asset/installconfig/powervs/client.go:534]: That check came from IBM Cloud code, and turns out it's rather misleading (giving the idea of potentially skipping), and in fact, it virtually never happens. So, I'm dropping the check in the next force-push. Thanks!!

**Comment by miyamotoh** [comment]
All checks passed. @barbacbd PTAL.

File: pkg/asset/installconfig/powervs/validation.go [hunk 4]
```diff
@@ -109,3 +110,75 @@ func validatePreExistingPrivateDNS(fldPath *field.Path, client API, ic *types.In
...
+		}
+	}
+	if !found {
+		allErrs = append(allErrs, field.NotFound(path.Child("vpcName"), name))
+	}
+
+	return allErrs
+}
+
+func findSubnetinVPC(client API, subnets []string, region string, name string, path *field.Path) field.ErrorList {
```
Comment (barbacbd) [inline pkg/asset/installconfig/powervs/validation.go:169]: nit: The i in In should be capitalized

File: pkg/asset/installconfig/powervs/client.go [hunk 5]
```diff
@@ -515,3 +516,22 @@ func (c *Client) GetAuthenticatorAPIKeyDetails(ctx context.Context) (*iamidentit
...
+
+	err := c.SetVPCServiceURLForRegion(ctx, region)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to set vpc api service url")
+	}
+
+	allVPCs := []vpcv1.VPC{}
+	if vpcs, _, err := c.vpcAPI.ListVpcs(c.vpcAPI.NewListVpcsOptions()); err != nil {
+		return nil, err
+	} else if vpcs != nil {
```
Comment (barbacbd) [inline pkg/asset/installconfig/powervs/client.go:533]: technically this else if is not needed.

File: pkg/asset/cluster/tfvars.go [hunk 6]
```diff
@@ -819,6 +824,46 @@ func (t *TerraformVariables) Generate(parents asset.Parents) error {
...
+		}
+		if vpcSubnet != "" {
+			var sn *vpcv1.Subnet
+			sn, err = client.GetSubnetByName(ctx, vpcSubnet, vpcRegion)
+			if err != nil {
+				return err
+			}
+			vpcZone = *sn.Zone.Name
+		} else {
+			rand.Seed(time.Now().UnixNano())
```
Comment (barbacbd) [inline pkg/asset/cluster/tfvars.go:863]: I am not familiar with powervs but looking at this I see some potentials for errors. What if there are other options for sub nets. Should we search for valid subnets and if you still want to use a random, randomly select from the list of valid subnets ?

Comment (clnperez) [reply pkg/asset/cluster/tfvars.go:863]: this is the way we've done it for a while. we (@mjturek and I) made a lot of time making sure it wasn't going to cause problems. however, we know it's not ideal not randomly select the vpc zone this way. but it's fine for now :)

Comment (miyamotoh) [reply pkg/asset/cluster/tfvars.go:863]: Not officially documented, but this knowledge of "valid zones" with the number 1 to 3 as suffix per region is gained from internal discussion with the IBM Cloud team. Therefore, this bit is basically allowing user to reuse his/her existing subnet, making sure it exists and deriving the zone from it, if specified, or we just pick one randomly from existing ones for the installer.

File: pkg/asset/installconfig/powervs/validation.go (hunk 4 above)
Comment (miyamotoh) [reply pkg/asset/installconfig/powervs/validation.go]: Addressed in the latest force-push.

File: pkg/asset/installconfig/powervs/client.go (hunk 5 above)
Comment (miyamotoh) [reply pkg/asset/installconfig/powervs/client.go]: Addressed in the latest force-push.

**Review by barbacbd — COMMENTED** [review]
/lgtm

File: pkg/asset/cluster/tfvars.go (hunk 6 above)
Comment (barbacbd) [reply pkg/asset/cluster/tfvars.go:863]: thanks for the explanations !

**Comment by r4f4** [comment]
/override ci/prow/e2e-azure-ovn
Install succeeded. Timed out during test after 4h.

**Comment by r4f4** [comment]
/override ci/prow/golint
Ran the linter locally with a longer timeout, no problems reported.

File: pkg/asset/installconfig/powervs/client.go [hunk 7]
```diff
@@ -516,3 +517,21 @@ func (c *Client) GetAuthenticatorAPIKeyDetails(ctx context.Context) (*iamidentit
...
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to set vpc api service url")
+	}
+
+	allVPCs := []vpcv1.VPC{}
+	if vpcs, _, err := c.vpcAPI.ListVpcs(c.vpcAPI.NewListVpcsOptions()); err != nil {
+		return nil, err
+	}
+	allVPCs = append(allVPCs, vpcs.Vpcs...)
+	return allVPCs, nil
```
Comment (r4f4) [inline pkg/asset/installconfig/powervs/client.go:536]: pkg/asset/installconfig/powervs/client.go:532:5: vpcs declared but not used
pkg/asset/installconfig/powervs/client.go:535:28: undefined: vpcs 

`vpcs` does not live beyound the scope of `if`.

```suggestion
	vpcs, _, err := c.vpcAPI.ListVpcs(c.vpcAPI.NewListVpcsOptions())
	if err != nil {
		return nil, err
	}
	return vpcs.Vpcs, nil
```

Comment (miyamotoh) [reply pkg/asset/installconfig/powervs/client.go]: Whoa, sorry about that. I thought I had force-pushed my latest commit, but apparently not. With that now pushed, it passes `go-lint`, `go-fmt`, and `go-vet` (tested locally). Let's see...

File: pkg/asset/installconfig/powervs/client.go [hunk 8]
```diff
@@ -516,3 +517,23 @@ func (c *Client) GetAuthenticatorAPIKeyDetails(ctx context.Context) (*iamidentit
...
+	}
+
+	vpcs, _, err := c.vpcAPI.ListVpcs(c.vpcAPI.NewListVpcsOptions())
+	if err != nil {
+		return nil, err
+	}
+
+	allVPCs := []vpcv1.VPC{}
+	allVPCs = append(allVPCs, vpcs.Vpcs...)
+	return allVPCs, nil
```
Comment (r4f4) [inline pkg/asset/installconfig/powervs/client.go:538]:
```suggestion
	return vpcs.Vpcs, nil
```
Any reason to create a new slice instead of returning the existing one?

Comment (miyamotoh) [reply pkg/asset/installconfig/powervs/client.go:538]: Will be addressed in the next force-push.

File: pkg/asset/cluster/tfvars.go (hunk 1 above)
Comment (r4f4) [reply pkg/asset/cluster/tfvars.go:69]: It seems the double import was brought back.

File: pkg/asset/cluster/tfvars.go [hunk 9]
```diff
@@ -819,6 +824,46 @@ func (t *TerraformVariables) Generate(parents asset.Parents) error {
...
+			vpcZone = fmt.Sprintf("%s-%d", vpcRegion, rand.Intn(2)+1)
```
Comment (r4f4) [inline pkg/asset/cluster/tfvars.go:864]:
```suggestion
			vpcZone = fmt.Sprintf("%s-%d", vpcRegion, rand.Intn(2)+1) //nolint:gosec // we don't need a crypto secure number
```

File: pkg/asset/installconfig/powervs/client.go [hunk 10]
```diff
@@ -516,3 +517,23 @@ func (c *Client) GetAuthenticatorAPIKeyDetails(ctx context.Context) (*iamidentit
 func (c *Client) GetAPIKey() string {
 	return c.APIKey
 }
+
+// GetVPCs gets all VPCs in a region
```
Comment (r4f4) [inline pkg/asset/installconfig/powervs/client.go:521]:
```suggestion
// GetVPCs gets all VPCs in a region.
```

File: pkg/asset/installconfig/powervs/validation.go [hunk 11]
```diff
@@ -110,3 +111,75 @@ func validatePreExistingPrivateDNS(fldPath *field.Path, client API, ic *types.In
 	}
 	return allErrs
 }
+
+// ValidateCustomVPCSetup ensures optional VPC settings, if specified, are all legit
```
Comment (r4f4) [inline pkg/asset/installconfig/powervs/validation.go:115]:
```suggestion
// ValidateCustomVPCSetup ensures optional VPC settings, if specified, are all legit.
```

File: pkg/types/powervs/powervs_regions.go [hunk 12]
```diff
@@ -97,3 +97,15 @@ func RegionShortNames() []string {
 	}
 	return keys
 }
+
+// ValidateVPCRegion validates that given VPC region is known/tested
```
Comment (r4f4) [inline pkg/types/powervs/powervs_regions.go:101]:
```suggestion
// ValidateVPCRegion validates that given VPC region is known/tested.
```

File: pkg/types/powervs/platform.go [hunk 13]
```diff
@@ -17,15 +17,15 @@ type Platform struct {
...
+
 	// VPCRegion specifies the IBM Cloud region in which to create VPC resources.
 	// Leave unset to allow installer to select the closest VPC region.
 	//
 	// +optional
 	VPCRegion string `json:"vpcRegion,omitempty"`
 
-	// UserID is the login for the user's IBM Cloud account.
-	UserID string `json:"userID"`
-
```
Comment (r4f4) [inline pkg/types/powervs/platform.go:28]: Any reason for this change? If so, it should go in its own commit.

File: pkg/asset/cluster/tfvars.go (hunk 1 above)
Comment (r4f4) [inline pkg/asset/cluster/tfvars.go:69]: Linter is still complaining about this double import.

Comment (r4f4) [reply pkg/asset/cluster/tfvars.go:68]: There was a recent commit adding linters to the config, so you might not have it if you haven't rebased on top of master recently

Comment (r4f4) [reply pkg/asset/cluster/tfvars.go:68]: Instead of removing the previous import, I'd suggest removing the one you're introducing and replacing `powervstypes` by `powervs` in your code. The build is broken right now.

**Comment by r4f4** [comment]
/override ci/prow/e2e-azure-ovn
Install and e2e succeeded, timed out during gather

**Comment by r4f4** [comment]
/override ci/prow/e2e-azure-ovn ci/prow/e2e-gcp-ovn
Install succeeded, timed out after 4h.
/hold cancel

**Comment by r4f4** [comment]
/override ci/prow/e2e-vsphere-ovn
Unrelated e2e failures
